From d2b083cf3ee74952802f3d8457ee7e24173ba303 Mon Sep 17 00:00:00 2001
From: Yen Chi Hsuan <yan12125@gmail.com>
Date: Wed, 23 Aug 2017 22:39:21 +0800
Subject: [PATCH] Link mozjs libraries correctly

As described in Mozilla bugs [1][2], linking to mozjs requires
additional flags or there will be segmentation faults. Specifically,
explicit linking to libmozglue.a is necessary. To prevent duplicated
symbols, mozjs should be added to individual components instead of put
into common packages.

Requires patching spidermonkey [3]

[1] https://bugzilla.mozilla.org/show_bug.cgi?id=1176787
[2] https://bugzilla.mozilla.org/show_bug.cgi?id=1236085
[3] https://aur.archlinux.org/cgit/aur.git/tree/link-mozglue.patch?h=js52
---
 Makefile-modules.am | 3 ++-
 Makefile-test.am    | 2 ++
 Makefile.am         | 5 ++++-
 configure.ac        | 5 +++--
 4 files changed, 11 insertions(+), 4 deletions(-)

diff --git a/Makefile-modules.am b/Makefile-modules.am
index 23c9d1bd..ee42aca7 100644
--- a/Makefile-modules.am
+++ b/Makefile-modules.am
@@ -11,7 +11,8 @@ libgjs_la_LIBADD += $(NATIVE_MODULES)
 JS_NATIVE_MODULE_CPPFLAGS =	\
 	$(AM_CPPFLAGS)		\
 	-DGJS_COMPILATION	\
-	$(GJS_CFLAGS)
+	$(GJS_CFLAGS)		\
+	$(MOZJS_CFLAGS)
 JS_NATIVE_MODULE_LIBADD =	\
 	$(GJS_LIBS)
 
diff --git a/Makefile-test.am b/Makefile-test.am
index 9cbdbe83..877d09f6 100644
--- a/Makefile-test.am
+++ b/Makefile-test.am
@@ -87,6 +87,7 @@ gjs_tests_CPPFLAGS =				\
 	$(AM_CPPFLAGS)				\
 	-DGJS_COMPILATION			\
 	$(GJSTESTS_CFLAGS)			\
+	$(MOZJS_CFLAGS)				\
 	$(gjs_directory_defines)		\
 	-I$(top_srcdir)/test
 
@@ -118,6 +119,7 @@ minijasmine_SOURCES =			\
 minijasmine_CPPFLAGS =				\
 	$(AM_CPPFLAGS)				\
 	$(GJS_CFLAGS)				\
+	$(MOZJS_CFLAGS)				\
 	-I$(top_srcdir)				\
 	-DINSTTESTDIR=\"$(gjsinsttestdir)\"	\
 	-DPKGLIBDIR=\"$(pkglibdir)\"		\
diff --git a/Makefile.am b/Makefile.am
index 38d0406e..773c3a20 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -56,6 +56,7 @@ lib_LTLIBRARIES += libgjs.la
 libgjs_la_CPPFLAGS =		\
 	$(AM_CPPFLAGS)		\
 	$(GJS_CFLAGS)	\
+	$(MOZJS_CFLAGS)		\
 	$(gjs_directory_defines)\
 	-I$(top_srcdir)/gi	\
 	-DGJS_COMPILATION
@@ -64,7 +65,8 @@ libgjs_la_LDFLAGS = 		\
 	-no-undefined \
 	$(NULL)
 libgjs_la_LIBADD = 		\
-	$(GJS_LIBS)
+	$(GJS_LIBS)		\
+	$(MOZJS_LIBS)
 
 if ENABLE_GTK
 libgjs_la_CPPFLAGS += $(GJS_GTK_CFLAGS)
@@ -137,6 +139,7 @@ bin_PROGRAMS += gjs-console
 gjs_console_CPPFLAGS = 		\
 	$(AM_CPPFLAGS)		\
 	$(GJS_CONSOLE_CFLAGS)	\
+	$(MOZJS_CFLAGS)		\
 	$(NULL)
 gjs_console_LDADD =		\
 	$(GJS_CONSOLE_LIBS)	\
diff --git a/configure.ac b/configure.ac
index 4e8428d8..df0342c8 100644
--- a/configure.ac
+++ b/configure.ac
@@ -59,7 +59,7 @@ GOBJECT_INTROSPECTION_REQUIRE([1.41.4])
 
 GOBJECT_REQUIREMENT="gobject-2.0 >= glib_required_version"
 gjs_base_packages="$GOBJECT_REQUIREMENT gio-2.0"
-common_packages="gthread-2.0 gio-2.0 >= glib_required_version mozjs-52"
+common_packages="gthread-2.0 gio-2.0 >= glib_required_version"
 gjs_packages="gobject-introspection-1.0 libffi $common_packages"
 gjs_cairo_packages="cairo cairo-gobject $common_packages"
 gjs_gtk_packages="gtk+-3.0 >= 3.20"
@@ -71,6 +71,7 @@ dnl These don't need to be put in the .pc file so use regular PKG_CHECK_MODULES
 PKG_CHECK_MODULES([GJS_GDBUS], [$gjs_base_packages])
 PKG_CHECK_MODULES([GJS_CONSOLE], [$gjs_base_packages])
 PKG_CHECK_MODULES([GJSTESTS], [$gjstests_packages])
+PKG_CHECK_MODULES([MOZJS], [mozjs-52])
 
 # Optional cairo dep (enabled by default)
 AC_ARG_WITH(cairo,
@@ -191,7 +192,7 @@ dnl If SpiderMonkey was compiled with --enable-debug, then we need to define
 dnl -DDEBUG before including js-config.h.
 AC_MSG_CHECKING([whether SpiderMonkey was configured with --enable-debug])
 CPPFLAGS_save="$CPPFLAGS"
-CPPFLAGS="$GJS_CFLAGS"
+CPPFLAGS="$MOZJS_CFLAGS"
 AC_COMPILE_IFELSE([AC_LANG_SOURCE([[
 #include <js-config.h>
 #ifdef JS_DEBUG
